### PR TITLE
View rich diffs for Source property changes

### PR DESCRIPTION
**Pull request by boatbomber**
The property inspection on .Source changes is just not helpful in the current state:

image

To remedy this, I've implemented [Neil Fraser's diffing](https://github.com/google/diff-match-patch) in Luau and used it to create meaningful Source diff pop-out widgets. It uses my open source [Highlighter](https://github.com/boatbomber/Highlighter) module to apply syntax highlighting that matches the user's Studio preferences.

image
image

### CODE REVIEW

File: plugin/src/App/StatusPages/Connected.lua [hunk 1]
```diff
@@ -367,13 +378,59 @@ function ConnectedPage:render()
...
 				end,
 			}),
+
+			SourceDiff = e(StudioPluginGui, {
+				id = "Rojo_ConnectedSourceDiff",
+				title = "Source diff",
+				active = self.state.showingSourceDiff,
+
+				initDockState = Enum.InitialDockState.Float,
+				initEnabled = true,
```
Comment (chriscerie) [inline plugin/src/App/StatusPages/Connected.lua:403]: What does this prop do? This looks unused.

File: plugin/src/App/StatusPages/Confirming.lua [hunk 2]
```diff
@@ -120,6 +135,44 @@ function ConfirmingPage:render()
...
 				PaddingRight = UDim.new(0, 20),
 			}),
+
+			SourceDiff = e(StudioPluginGui, {
+				id = "Rojo_ConfirmingSourceDiff",
+				title = "Source diff",
+				active = self.state.showingSourceDiff,
+
+				initDockState = Enum.InitialDockState.Float,
+				initEnabled = true,
```
Comment (chriscerie) [inline plugin/src/App/StatusPages/Confirming.lua:145]: Same thing here with unused prop

File: plugin/src/App/Components/StringDiffVisualizer/init.lua [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+local e = Roact.createElement
+
+local StringDiffVisualizer = Roact.Component:extend("StringDiffVisualizer")
+
+function StringDiffVisualizer:init()
+	self.scriptBackground, self.setScriptBackground = Roact.createBinding(Color3.fromRGB(0, 0, 0))
+	self.contentSize, self.setContentSize = Roact.createBinding(Vector2.new(0, 0))
+
+	self:calculateContentSize()
+	self:getScriptBackground()
```
Comment (chriscerie) [inline plugin/src/App/Components/StringDiffVisualizer/init.lua:27]: Can we rename this function to be something that better indicates it performs a side effect?

File: plugin/src/App/Components/StringDiffVisualizer/init.lua [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+	self:setState({
+		add = {},
+		remove = {},
+	})
+end
+
+function StringDiffVisualizer:getScriptBackground()
+	self.setScriptBackground(Highlighter.getTokenColor("background"))
```
Comment (chriscerie) [inline plugin/src/App/Components/StringDiffVisualizer/init.lua:36]: Can we get a comment here explaining why we are setting background color to the "background" token on every render despite it being the only place that sets the binding

**Review by chriscerie — COMMENTED** [review]
Great work

File: plugin/src/App/StatusPages/Connected.lua (hunk 1 above)
Comment (boatbomber) [reply plugin/src/App/StatusPages/Connected.lua:403]: Huh. You're right. I copied it from App, I guess it did something once upon a time and was never cleaned up. Good eye!

File: plugin/src/App/Components/CodeLabel.lua [hunk 5]
```diff
@@ -0,0 +1,62 @@
+local Rojo = script:FindFirstAncestor("Rojo")
+local Plugin = Rojo.Plugin
```
Comment (Dekkonot) [inline plugin/src/App/Components/CodeLabel.lua:2]:
```suggestion
```

This variable isn't used

**Review by Dekkonot — CHANGES_REQUESTED** [review]

**Review by Dekkonot — APPROVED** [review]

**Pull request merged**